tikv/tikv
# s3: support backup with session token and assume role

### What is changed and how it works?

Issue Number: Close https://github.com/pingcap/tidb/issues/39832
Issue Number: Close https://github.com/tikv/tikv/issues/15781

What's Changed:
1. support retrive `session-token` from input. 
2. support assume role with `role-arn` and `extrenal-id` and keep the same behavior with tidb side.

```  
 s3: support backup with session token and assume role
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

session-token:
1. set env of token.
```
export AWS_ACCESS_KEY_ID="xxx"
export AWS_SECRET_ACCESS_KEY="yyy"
export AWS_SESSION_TOKEN="zzz"
```
2. do a full backup and check backup data in s3.
<img width="1506" alt="image" src="https://github.com/tikv/tikv/assets/5906259/e748dfe1-5cf4-4690-a6b5-4782f989dd9a">

role-arn:
1. prepare two AWS account (A, B).
2. create a bucket in account A.
3. create a role for this bucket in account A.
4. create a trust relationship in account A for account B to assume above role.
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::ACCOUNTB:root"
            },
            "Action": "sts:AssumeRole",
            "Condition": {
                "StringEquals": {
                    "sts:ExternalId": "test-external-id"
                }
            }
        }
    ]
}
```

run backup with role-arn and external-id.

<img width="758" alt="image" src="https://github.com/tikv/tikv/assets/5906259/b8b3f3ad-8e8f-46f6-9ec5-0213a662ccf4">


Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
support backup and restore with session token and assume role.
```
